frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseAction.java

```diff
@@ -0,0 +1,287 @@
+import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+import com.virtuslab.qual.gitmachete.backend.api.ConfirmedNonRoot;
+
+@ExtensionMethod({GitVfsUtils.class, GitMacheteBundle.class})
+@CustomLog
+public class TraverseAction extends BaseGitMacheteRepositoryReadyAction implements IExpectsKeySelectedBranchName {
+
+  @Untainted
+  @I18nFormat({I18nConversionCategory.GENERAL})
+  final String actionIdFormatString = getNonHtmlString("action.GitMachete.TraverseAction.invoked-action.id");
+
+  @Override
+  public LambdaLogger log() {
+    return LOG;
+  }
+
+  public static final String SHOW_PULL_APPROVAL = "git-machete.pull.approval.show";
+  public static final String SHOW_PUSH_APPROVAL = "git-machete.push.approval.show";
+  public static final String SHOW_TRAVERSE_APPROVAL = "git-machete.traverse.approval.show";
+
+  @Override
+  @UIEffect
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+    presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseAction.description.text"));
+
+    val repositorySnapshot = getGitMacheteRepositorySnapshot(anActionEvent);
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+
+    if (branchLayout == null || branchLayout.getRootEntries().isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.TraverseAction.description.empty-layout"));
+      return;
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val repository = getSelectedGitRepository(anActionEvent);
+    val branchLayout = getBranchLayout(anActionEvent);
+    val project = getProject(anActionEvent);
+
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && repository != null) {
+      boolean traverseApproved = true;
+      if (PropertiesComponent.getInstance(project).getBoolean(SHOW_TRAVERSE_APPROVAL, /* defaultValue */ true)) {
+
+        val traverseApprovalDialogBuilder = MessageDialogBuilder.okCancel(
+            getString("action.GitMachete.TraverseAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.TraverseAction.dialog.traverse-approval.text.HTML"));
+
+        traverseApprovalDialogBuilder
+            .yesText(getString("action.GitMachete.TraverseAction.dialog.traverse-approval.ok-text"))
+            .icon(Messages.getQuestionIcon()).doNotAsk(new TraverseApprovalDialog(project));
+
+        traverseApproved = traverseApprovalDialogBuilder.ask(project);
+      }
+
+      if (traverseApproved) {
+        var currentBranchName = getCurrentBranchNameIfManaged(anActionEvent);
```

Comment:
also for now having checked out an unmanaged branch and running traverse once I agree in the traverse dialog (yes, traverse) nothing happens then. 

shouldn't the traverse simply checkout first root and start from there no matter what the current branch is?

Reply:
There is an issue with synchronizing the different stages, which would be solved by `Task.Backgroundable` and `onSuccess`